storm-core/src/jvm/org/apache/storm/executor/spout/SpoutExecutor.java

```diff
@@ -0,0 +1,243 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.storm.executor.spout;
+
+import com.google.common.collect.ImmutableMap;
+import org.apache.storm.Config;
+import org.apache.storm.Constants;
+import org.apache.storm.ICredentialsListener;
+import org.apache.storm.daemon.Acker;
+import org.apache.storm.daemon.StormCommon;
+import org.apache.storm.daemon.Task;
+import org.apache.storm.daemon.metrics.BuiltinMetricsUtil;
+import org.apache.storm.daemon.metrics.SpoutThrottlingMetrics;
+import org.apache.storm.executor.Executor;
+import org.apache.storm.executor.TupleInfo;
+import org.apache.storm.hooks.info.SpoutAckInfo;
+import org.apache.storm.hooks.info.SpoutFailInfo;
+import org.apache.storm.spout.ISpout;
+import org.apache.storm.spout.ISpoutWaitStrategy;
+import org.apache.storm.spout.SpoutOutputCollector;
+import org.apache.storm.stats.SpoutExecutorStats;
+import org.apache.storm.tuple.TupleImpl;
+import org.apache.storm.utils.*;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+public class SpoutExecutor extends Executor {
+
+    private static final Logger LOG = LoggerFactory.getLogger(SpoutExecutor.class);
+
+    private final ISpoutWaitStrategy spoutWaitStrategy;
+    private Integer maxSpoutPending;
+    private final AtomicBoolean lastActive;
+    private List<ISpout> spouts;
+    private List<SpoutOutputCollector> outputCollectors;
+    private final MutableLong emittedCount;
+    private final MutableLong emptyEmitStreak;
+    private final SpoutThrottlingMetrics spoutThrottlingMetrics;
+    private final boolean hasAckers;
+    private RotatingMap<Long, TupleInfo> pending;
+    private final boolean backPressureEnabled;
+
+    public SpoutExecutor(final Map workerData, final List<Long> executorId, Map<String, String> credentials) {
+        super(workerData, executorId, credentials);
+        this.spoutWaitStrategy = Utils.newInstance((String) stormConf.get(Config.TOPOLOGY_SPOUT_WAIT_STRATEGY));
+        this.spoutWaitStrategy.prepare(stormConf);
+
+        this.backPressureEnabled = Utils.getBoolean(stormConf.get(Config.TOPOLOGY_BACKPRESSURE_ENABLE), false);
+
+        this.lastActive = new AtomicBoolean(false);
+        this.hasAckers = StormCommon.hasAckers(stormConf);
+        this.emittedCount = new MutableLong(0);
+        this.emptyEmitStreak = new MutableLong(0);
+        this.spoutThrottlingMetrics = new SpoutThrottlingMetrics();
+    }
+
+    @Override
+    public void init(final Map<Integer, Task> idToTask) {
+        LOG.info("Opening spout {}:{}", componentId, idToTask.keySet());
+        this.idToTask = idToTask;
+        this.maxSpoutPending = Utils.getInt(stormConf.get(Config.TOPOLOGY_MAX_SPOUT_PENDING), 0) * idToTask.size();
```

Comment:
The default value of max pending is null, in the original executor.clj:
```
^Integer max-spout-pending (if max-spout-pending (int max-spout-pending))
```
max-spout-pending could be null

thus in the code usage, it's like `max-spout-pending and ...`, while I changed it into java in this way:
`maxSpoutPending !=0 && ...`